external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    diag,
+    diagnostics::codes::TypeSafety,
+    expansion::ast::{self as E, ModuleIdent},
+    naming::ast::{self as N, StructFields, Type, Type_},
+    parser::ast::{ConstantName, DatatypeName, FunctionName},
+    shared::{
+        known_attributes::{AttributePosition, KnownAttribute},
+        CompilationEnv, Identifier, Name,
+    },
+    typing::{ast as T, visitor::TypingVisitorContext},
+};
+use move_ir_types::location::*;
+use move_proc_macros::growing_stack;
+use std::collections::{BTreeMap, BTreeSet};
+
+const NOTE_STR: &str = "note";
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct Deprecation {
+    // The source location of the deprecation attribute
+    source_location: Loc,
+    // The type of the member that is deprecated (function, constant, etc.)
+    location: AttributePosition,
```

Comment:
Do you need this here? Isn't it recoverable by knowing what `deprecated_*` entry you are looking at?

Reply:
Nope! Because of module deprecations. If we want to be able to disambiguate function/struct/const members in the warning messages then we need to pass this through.